moshi/src/main/java/com/squareup/moshi/BufferedSourceJsonReader.java

```diff
@@ -718,15 +719,15 @@ private boolean isLiteral(int c) throws IOException {
     }
 
     peeked = PEEKED_BUFFERED;
-    double asDouble;
+    BigDecimal asDecimal;
     try {
-      asDouble = Double.parseDouble(peekedString);
+      asDecimal = new BigDecimal(peekedString);
     } catch (NumberFormatException e) {
       throw new JsonDataException("Expected a long but was " + peekedString
           + " at path " + getPath());
     }
-    long result = (long) asDouble;
-    if (result != asDouble) { // Make sure no precision was lost casting to 'long'.
+    long result = asDecimal.longValue();
+    if (result != asDecimal.doubleValue()) { // Make sure no precision was lost casting to 'long'.
```

Comment:
Ok, the limit check breaks... If the parsed long is bigger then MAX_VALUE or smaller then MIN_VALUE this ensure validity. 
Trying to convert the result into a BigDecemal will break the comparison for some valid use cases.